### PR TITLE
Feature: custom user color and alias

### PR DESCRIPTION
**Pull request by rod41732**
Allows defining custom color and/or alias for chatters

Screenshots
- How aliased users looks like in Chat
[Screenshot_20221123-150917](https://user-images.githubusercontent.com/26536811/203499209-d9410bc7-b93a-46fb-aa63-1791b337fe01.png)
- Settings Screen (Settings -> Chat -> Custom User Display) (target user, color, alias)
[Screenshot_20221123-150954](https://user-images.githubusercontent.com/26536811/203499217-1c0cd4d1-c021-45c7-aa92-d67d5a6bb222.png)

Note:
- name matching is case-insensitive
- leaving blank in color or alias will not override their color or name
- color needs to be prefixed with \# (it's parsed by `Color.parseColor`). Entering invalid color simply has no effect
- change take place "immediately" for new messages after settings applied, but old message won't be affected. Essentially, restart would be required.

### CODE REVIEW

**Comment by flex3r** [comment]
Some comments regarding the ui:
image

File: app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayDao.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+@Dao
+interface UserDisplayDao {
+    @Query("SELECT * from user_display")
+    suspend fun getAll(): List<UserDisplayEntity>
+
+    @Query("SELECT * from user_display")
+    fun getUserDisplaysFlow(): Flow<List<UserDisplayEntity>>
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayDao.kt:14]: Use `@Upsert`

File: app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayEntity.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    @ColumnInfo(name = "target_user") val targetUser: String, // target user to apply the color and/or alias
+    @ColumnInfo(name = "enabled") val enabled: Boolean,
+    @ColumnInfo(name = "color_enabled") val colorEnabled: Boolean,
+    @ColumnInfo(name = "color_hex") val color: Int,
+    @ColumnInfo(name = "aliasEnabled") val aliasEnabled: Boolean,
+    @ColumnInfo(name = "alias") val alias: String? // aliased name
+) {
+    companion object {
+        /** make dummy entity with specified ID, useful for deleting by ID, values here doesn't matter at all*/
+        fun makeDummy(id: Int) = UserDisplayEntity(id = id, targetUser = "", enabled = false, colorEnabled = false, color = 0, aliasEnabled = false, alias = "")
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayEntity.kt:19]: Why is this needed?

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+package com.flxrs.dankchat.preferences.userdisplay
+
+import androidx.annotation.ColorInt
+import com.flxrs.dankchat.data.database.UserDisplayEntity
+import kotlinx.serialization.Serializable
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+
+@Serializable
+data class UserDisplayDto(
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt:12]: This file serves no real purpose at the moment. It is located in the ui package but is also part of some of the messages, which are located in the data package.
I would instead like the approach like I did for Highlights/Ignores. I have some kind of model that has only the properties needed for the messages, like the class `Highlight` this would be located in the different Message classes.
For the bottom sheet, the `UserDisplayViewModel` can directly consume the database entities and map them to the `UserDisplayItem`, skipping this dto mapping completely.

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt [hunk 4]
```diff
@@ -124,4 +146,55 @@ class ChatSettingsFragment : MaterialPreferenceFragmentCompat() {
 
         return true
     }
+
+    private fun showUserDisplaySettingsFragment(root: View): Boolean {
+        val context = root.context
+        val windowHeight = resources.displayMetrics.heightPixels
+        val peekHeight = (windowHeight * 0.6).roundToInt()
+
+        lifecycleScope.launch {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:155]: Create a UserDisplayViewModel and move all of this logic into there please

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        color = color,
+        aliasEnabled = aliasEnabled,
+        alias = alias
+    )
+}
+
+// these are defined as extension type to allow convenient checking even the object is null
+/** return whether alias is set, and enabled
+ * calling on null will return false */
+fun UserDisplayDto?.hasAlias(): Boolean {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt:68]: I would like to see some of this validation logic handled by the repository, if possible, similar to how I only add valid highlights/ignores to messages. The ui should just get a static "state" which already contains everything needed.

File: app/src/main/kotlin/com/flxrs/dankchat/utils/extensions/ColorExtensions.kt [hunk 6]
```diff
@@ -44,4 +45,22 @@ fun Int.normalizeColor(@ColorInt background: Int): Int {
...
+/** convert int to RGB with zero pad */
+fun Int.toHexCode(): String = Integer.toHexString(this.onlyRGB()).padStart(6, '0')
+
+/** convert RGB color (0xffffff) to ARGB with alpha */
+fun Int.toARGBInt(alpha: Int = 255): Int = (alpha shl 24) or this.onlyRGB()
+
+
+/** find best contrast text to display on specified background color (whtie or black) -- opacity ignored */
+fun Int.getContrastTextColor(): Int {
+    val possibleColors = listOf(Color.BLACK, Color.WHITE)
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/utils/extensions/ColorExtensions.kt:63]: Keep in mind that the regular text colors are not black or white, instead `R.attr.colorOnSurface` is used , depending on the current theme and the inverse `colorOnSurface`. So I would instead like to see those colors being used for the current theme.
Also instead of this hacky approach you can probably just use `MaterialColors.isColorLight`

File: app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayEntity.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "user_display")
+data class UserDisplayEntity(
+    @PrimaryKey(autoGenerate = true) val id: Int,
+    @ColumnInfo(name = "target_user") val targetUser: String, // target user to apply the color and/or alias
+    @ColumnInfo(name = "enabled") val enabled: Boolean,
+    @ColumnInfo(name = "color_enabled") val colorEnabled: Boolean,
+    @ColumnInfo(name = "color_hex") val color: Int,
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/database/UserDisplayEntity.kt:13]: Don't save it as a plain hex but instead save it as a `Color` type and add a custom converter instead of manually doing the conversion

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt (hunk 4 above)
Comment (flex3r) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:155]: A UI component should never directly communicate with a component in the data layer, instead it should be delegated to a state holder, in this case a viewholder. See also https://developer.android.com/topic/architecture#ui-layer

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        emptyList()
+    )
+
+    /** add (if ID == 0) or update (ID != 0) user displays */
+    suspend fun addUserDisplays(userDisplays: List<UserDisplayDto>) {
+        userDisplayDao.insertAll(userDisplays.map { it.toEntity() })
+    }
+
+    suspend fun deleteByIds(ids: List<Int>): Int {
+        return userDisplayDao.deleteAll(ids.map { UserDisplayEntity.makeDummy(it) })
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:36]: This is not needed if you go with the approach I use for highlights/ignores. I just map the `Items` used by the ui back to the corresponding entities.

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+) {
+    suspend fun getAllUserDisplays(): List<UserDisplayDto> = userDisplayDao.getAll().map { it.toDto() }
+
+    private val userDisplays = userDisplayDao.getUserDisplaysFlow().map { users -> users.map { it.toDto() } }.stateIn(
+        coroutineScope,
+        SharingStarted.Eagerly,
+        emptyList()
+    )
+
+    /** add (if ID == 0) or update (ID != 0) user displays */
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:30]: I don't understand this comment

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+import com.flxrs.dankchat.databinding.AddItemBinding
+import com.flxrs.dankchat.databinding.UserDisplayItemBinding
+import com.flxrs.dankchat.utils.extensions.getContrastTextColor
+import com.flxrs.dankchat.utils.extensions.toARGBInt
+import com.flxrs.dankchat.utils.extensions.toHexCode
+import com.google.android.material.button.MaterialButton
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+import com.rarepebble.colorpicker.ColorPickerView
+
+class UserDisplayAdapter(val entries: MutableList<UserDisplayItem>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:18]: Don't use the MutableList approach, use the ListAdapter approach with callbacks and introduce a ViewModel that does all the changes to the ui state. This way we can directly sync it with the db

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                entries.removeAt(bindingAdapterPosition)
+                notifyItemRemoved(bindingAdapterPosition)
+            }
+
+            binding.userDisplayPickColorButton.setOnClickListener {
+                val item = binding.userDisplay ?: return@setOnClickListener
+                val picker = ColorPickerView(binding.root.context)
+                picker.showAlpha(false)
+
+                picker.color = item.color.toARGBInt() // to correctly set initial alpha to 255
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:32]: This should be moved to the ViewMode, UI should not need to do this

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                    .setNegativeButton(binding.root.context.getString(R.string.dialog_cancel)) { _, _ -> }
+                    .setPositiveButton(binding.root.context.getString(R.string.dialog_ok)) { _, _ ->
+                        val pickedColor = picker.color
+                        item.color = pickedColor
+                        binding.userDisplayPickColorButton.setColorAndBg(item.color)
+                    }
+                    .show()
+            }
+
+            (binding.userDisplayEnableColor).apply {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:45]: Why these kind of applys when you only call `setOnCheckedChangeListener` inside?

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                    item.aliasEnabled = checked
+                    binding.userDisplayAliasInput.isVisible = checked
+                }
+            }
+        }
+    }
+
+    /** set text, text color, background color to represent specified color */
+    @SuppressLint("SetTextI18n")
+    private fun MaterialButton.setColorAndBg(colorRGB: Int) {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:65]: I would like to have separate properties in the `UserDisplayItem` UI state that contains the text, the textColor and the backgroundColor

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        text = "#" + colorRGB.toHexCode()
+        setTextColor(colorRGB.toARGBInt().getContrastTextColor())
+        setBackgroundColor(colorRGB.toARGBInt())
+    }
+
+    inner class AddItemViewHolder(val binding: AddItemBinding) : RecyclerView.ViewHolder(binding.root) {
+        init {
+            binding.multiEntryAdd.setOnClickListener {
+                // ID 0, so that the create call generate the ID
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:74]: Too much logic for the UI, should be delegated to the viewmodel which then calls the repository.

**Review by flex3r — CHANGES_REQUESTED** [review]
TLDR I'd suggest looking at the full highlights/ignore flow again, especially the data flow between UI, ViewModel (which is missing here rn) and Repository.
For the actual chat part I'd like a better separation with models that fit the Message classes better so logic inside the ChatAdapter can be reduced

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt (hunk 3 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt]: the `Dto` class will be removed, and `UserDisplayViewModel` will be implemented :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt (hunk 4 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:153]: sure :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt (hunk 5 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt]: I have gone with the way mentioned above (new data class `UserDisplayEffectiveValue`). There will similar methods on the class `UserDisplayEffectiveValue` tho, but the purpose is just to simplify null checking / handling (similar purpose like String's `ifEmpty`).

I also move the logic of formatting username / color to `Message` classes.

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt (hunk 10 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt]: gone with the way you suggested, much better code :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt (hunk 12 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:51]: Just me being dank when writing the code FeelsDankMan

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt (hunk 13 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:78]: sure :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/ClearChatMessage.kt [hunk 15]
```diff
@@ -33,4 +37,19 @@ data class ClearChatMessage(
...
         }
     }
-}
\ No newline at end of file
+
+
+}
+
+// Extracted from ChatAdapter, TODO localize
+val ClearChatMessage.fullText: String
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/ClearChatMessage.kt:45]: I would like these extensions to have more descriptive names. Here we could use something like `systemMessage`

**Review by flex3r — CHANGES_REQUESTED** [review]
I can't stop nit-picking, sorry. Overall very good changes and mainly small stuff left though LULE

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private fun PointRedemptionMessage.applyUserDisplay(): PointRedemptionMessage {
+        val match = findMatchingUserDisplay(name) ?: return this
+        return copy(userDisplay = match)
+    }
+
+    // e.g. announcement ->have child message
+    private fun UserNoticeMessage.applyUserDisplay(): UserNoticeMessage {
+        val processedChildMessage = childMessage?.applyUserDisplay()
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:75]: I want the to return immediately when `processedChildMessage` is same as `childMessage` to prevent unnecessary copying 

which means two scenarios:
- `childMessage` is null (and then `processedChildMessage` is also null)
- `childMessage` is not null, but after calling `applyUserDisplay`, it still return the same instance (i.e. return `this` as in the below code) 
```kotlin
        val match = findMatchingUserDisplay(name) ?: return this
```

the early you suggested return would only handle the first case tho.

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+            is PrivMessage            -> message.applyUserDisplay()
+            is UserNoticeMessage      -> message.applyUserDisplay()
+            is WhisperMessage         -> message.applyUserDisplay()
+            else                      -> return message
+        }
+
+    }
+
+    private fun findMatchingUserDisplay(name: String): UserDisplayEffectiveValue? {
+        return userDisplays.value.find { it.targetUser.equals(name, true) }?.toEntry()?.effectiveValue()
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:54]: won't that result in some duplicate code? :thinking: (FYI: Entry (`UserDisplayItem.Entry`) is almost one-to-one mapping from `Entity`, the difference is only some field name and being mutable vs immutable).

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+import android.content.Context
+import androidx.annotation.ColorInt
+import com.flxrs.dankchat.data.database.UserDisplayEntity
+import com.flxrs.dankchat.utils.extensions.getContrastTextColor
+import com.flxrs.dankchat.utils.extensions.toHexCode
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+
+data class UserDisplayEffectiveValue(val alias: String?, val color: Int?)
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt:13]: :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+fun UserDisplayEffectiveValue?.hasAlias(): Boolean {
+    contract {
+        returns(true) implies (this@hasAlias != null)
+    }
+    return this != null && alias != null
+}
+
+
+fun UserDisplayEffectiveValue?.nameOr(fallback: String): String {
+    return if (hasAlias()) alias!! else fallback
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt:74]: yeah I'll probably remove them, it's kind like left-over functions from before refactor.

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt [hunk 20]
```diff
@@ -0,0 +1,88 @@
+@file:OptIn(ExperimentalContracts::class)
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayItem.kt:1]: :+1:

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (flex3r) [reply app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:75]: Sounds like over-optimizing to me tbh LULE

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (flex3r) [reply app/src/main/kotlin/com/flxrs/dankchat/data/repo/UserDisplayRepository.kt:54]: Yes, you always have similar, duplicate looking code when mapping between ui, domain and entity models, that's a side-effect of this architecture. 
The separation of concerns is more important than some code duplication though

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt [hunk 21]
```diff
@@ -124,4 +143,40 @@ class ChatSettingsFragment : MaterialPreferenceFragmentCompat() {
...
+            val userDisplayAdapter = UserDisplayAdapter(
+                { currentEntries ->
+                    // ensure modified entries are saved, before new entry added
+                    userDisplayViewModel.saveEntries(currentEntries)
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:156]: I had problem when adding item item will cause changes in existing entry to be lost.
Probably thoroughly debug it again tho. (Was a quick fix to get things done)

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt [hunk 22]
```diff
@@ -81,4 +84,16 @@ val PrivMessage.isFirstMessage: Boolean
...
+    get() = tags["pinned-chat-paid-amount"] != null
+
+/** format name for display in chat */
+val PrivMessage.fullName: String
+    get() = userDisplay?.alias ?: when {
+        displayName.equals(name, ignoreCase = true) -> displayName
+        else                                        -> "$name($displayName)"
+    }
+
+fun PrivMessage.finalColor(@ColorInt bgColor: Int): Int {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt:96]: The naming here doesn't really tell what this function does, e.g. the fallback and color normalization.
Perhaps:
```suggestion
fun PrivMessage.customOrUserColor(@ColorInt background: Int): Int {
```

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt [hunk 23]
```diff
@@ -81,4 +84,16 @@ val PrivMessage.isFirstMessage: Boolean
     get() = tags["first-msg"] == "1"
 
 val PrivMessage.isElevatedMessage: Boolean
-    get() = tags["pinned-chat-paid-amount"] != null
\ No newline at end of file
+    get() = tags["pinned-chat-paid-amount"] != null
+
+/** format name for display in chat */
+val PrivMessage.fullName: String
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt:90]: I think something like `formattedName` fits better here.

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/WhisperMessage.kt [hunk 24]
```diff
@@ -88,4 +92,22 @@ data class WhisperMessage(
...
+}
+
+
+val WhisperMessage.senderFullName: String
+    get() = userDisplay?.alias ?: when {
+        displayName.equals(name, true) -> displayName
+        else                           -> "$name($displayName)"
+    }
+
+fun WhisperMessage.senderColorOnBg(@ColorInt bgColor: Int): Int = userDisplay.colorOrElse(color.normalizeColor(bgColor))
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/WhisperMessage.kt:105]:
```suggestion
fun WhisperMessage.senderColorOnBackground(@ColorInt background: Int): Int = userDisplay.colorOrElse(color.normalizeColor(background))
```

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt [hunk 25]
```diff
@@ -91,11 +113,7 @@ class ChatSettingsFragment : MaterialPreferenceFragmentCompat() {
         val windowHeight = resources.displayMetrics.heightPixels
         val peekHeight = (windowHeight * 0.6).roundToInt()
         val commands = runCatching {
-            sharedPreferences
-                .getStringSet(key, emptySet())
-                .orEmpty()
-                .mapNotNull { Json.decodeOrNull<CommandDto>(it)?.toEntryItem() }
-                .plus(CommandItem.AddEntry)
+            sharedPreferences.getStringSet(key, emptySet()).orEmpty().mapNotNull { Json.decodeOrNull<CommandDto>(it)?.toEntryItem() }.plus(CommandItem.AddEntry)
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:116]: Undo these changes please

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt [hunk 26]
```diff
@@ -109,11 +127,8 @@ class ChatSettingsFragment : MaterialPreferenceFragmentCompat() {
         bottomSheetDialog = BottomSheetDialog(context).apply {
             setContentView(binding.root)
             setOnDismissListener {
-                val stringSet = commandAdapter.commands
-                    .filterIsInstance<CommandItem.Entry>()
-                    .filter { it.trigger.isNotBlank() && it.command.isNotBlank() }
-                    .map { Json.encodeToString(it.toDto()) }
-                    .toSet()
+                val stringSet =
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:130]: Undo these changes please

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt [hunk 27]
```diff
@@ -124,4 +139,46 @@ class ChatSettingsFragment : MaterialPreferenceFragmentCompat() {
...
+                    height = windowHeight
+                }
+            }
+
+            collectFlow(userDisplayViewModel.userDisplays) { userDisplayAdapter.submitList(it) }
+            collectFlow(userDisplayViewModel.events) { event ->
+                when (event) {
+                    is UserDisplayEvent.ItemRemoved -> Snackbar.make(
+                        bottomSheetBinding.root, getString(R.string.item_removed), Snackbar.LENGTH_SHORT
+                    ).setAction("Undo") { userDisplayViewModel.saveChangesAndAddEntry(userDisplayAdapter.currentList, event.item) }.show()
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/ui/ChatSettingsFragment.kt:167]:
```suggestion
                    ).setAction(R.string.undo) { userDisplayViewModel.saveChangesAndAddEntry(userDisplayAdapter.currentList, event.item) }.show()
```

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+            binding.multiEntryAdd.setOnClickListener {
+                onAddItem(currentList)
+            }
+        }
+
+    }
+
+    /** set text, text color, background color to represent specified color */
+    @SuppressLint("SetTextI18n")
+    private fun MaterialButton.setColorAndBg(item: UserDisplayItem.Entry) {
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:72]:
```suggestion
    private fun MaterialButton.setColorAndBackground(item: UserDisplayItem.Entry) {
```

File: app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+                return oldItem.id == newItem.id
+            }
+            return oldItem == newItem
+        }
+
+        override fun areContentsTheSame(oldItem: UserDisplayItem, newItem: UserDisplayItem) = oldItem == newItem
+
+    }
+
+
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt:127]: More empty lines FeelsDankMan

File: app/src/main/kotlin/com/flxrs/dankchat/utils/extensions/Extensions.kt [hunk 30]
```diff
@@ -15,15 +15,10 @@ import com.google.android.material.color.MaterialColors
...
-    ?.flatMap { it.value }
-    .orEmpty()
-
-fun List<GenericEmote>.moveToFront(channel: String?): List<GenericEmote> = this
-    .partition { it.emoteType.title.equals(channel, ignoreCase = true) }
-    .run { first + second }
+fun List<GenericEmote>?.toEmoteItems(): List<EmoteItem> =
+    this?.groupBy { it.emoteType.title }?.mapValues { (title, emotes) -> EmoteItem.Header(title) + emotes.map(EmoteItem::Emote).sorted() }?.flatMap { it.value }.orEmpty()
+
+fun List<GenericEmote>.moveToFront(channel: String?): List<GenericEmote> = this.partition { it.emoteType.title.equals(channel, ignoreCase = true) }.run { first + second }
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/utils/extensions/Extensions.kt:21]: More unrelated changes

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/UserDisplay.kt [hunk 31]
```diff
@@ -0,0 +1,27 @@
+package com.flxrs.dankchat.data.twitch.message
+
+import androidx.annotation.ColorInt
+import com.flxrs.dankchat.preferences.userdisplay.UserDisplayItem
+
+
+/** represent final effect UserDisplay (after considering enabled/disabled states) */
+data class UserDisplay(val alias: String?, val color: Int?) {
+    companion object {
+        fun UserDisplayItem.Entry.toEffectiveValue() = UserDisplay(
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/UserDisplay.kt:10]:
```suggestion
        fun UserDisplayItem.Entry.toUserDisplay() = UserDisplay(
```

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/WhisperMessage.kt [hunk 32]
```diff
@@ -88,4 +92,22 @@ data class WhisperMessage(
...
+
+val WhisperMessage.recipientFullName: String
+    get() = recipientDisplay?.alias ?: when {
+        recipientDisplayName.equals(recipientName, true) -> recipientDisplayName
+        else                                             -> "$recipientName($recipientDisplayName)"
+    }
+
+fun WhisperMessage.recipientColorOnBg(@ColorInt bgColor: Int): Int = recipientDisplay.colorOrElse(recipientColor.normalizeColor(bgColor))
```
Comment (flex3r) [inline app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/WhisperMessage.kt:113]:
```suggestion
fun WhisperMessage.recipientColorOnBackground(@ColorInt background: Int): Int = recipientDisplay.colorOrElse(recipientColor.normalizeColor(background))
```

**Review by flex3r — CHANGES_REQUESTED** [review]
0:13:32.139 doge41732: hardest thing in programming is naming Keepo

File: app/build.gradle [hunk 33]
```diff
@@ -62,8 +62,8 @@ android {
         }
         dank {
             initWith debug
-            minifyEnabled true
-            shrinkResources true
+            minifyEnabled false
+            shrinkResources false
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/build.gradle:66]: I intended to disable for me only, but accidentally committed FeelsDankMan. Enabled it again.

File: app/src/main/kotlin/com/flxrs/dankchat/data/repo/ChatRepository.kt [hunk 34]
```diff
@@ -466,7 +469,7 @@ class ChatRepository @Inject constructor(
 
     private fun handleClearChat(msg: IrcMessage) {
         val parsed = runCatching {
-            ClearChatMessage.parseClearChat(msg)
+            ClearChatMessage.parseClearChat(msg).calculateUserDisplays() as ClearChatMessage
```
[the remark this answers is not in the document]
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/data/repo/ChatRepository.kt:472]: I guess we should not, gonna remove it.

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt (hunk 22 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt:96]: I go with `customOrUserColorOn` to indicate more clearly than the passed color is background, not some kind of fallback value like "getOrElse" kind of functions.

File: app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt (hunk 23 above)
Comment (rod41732) [reply app/src/main/kotlin/com/flxrs/dankchat/data/twitch/message/PrivMessage.kt:90]: agreed :+1:

**Comment by rod41732** [comment]
LGTM :+1:

**Review by flex3r — APPROVED** [review]
Thanks for the awesome contribution image
